src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractNativeIntegerKeywordRecommender.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders
+{
+    internal abstract class AbstractNativeIntegerKeywordRecommender : IKeywordRecommender<CSharpSyntaxContext>
+    {
+        protected abstract RecommendedKeyword Keyword { get; }
+
+        private bool IsValidContext(CSharpSyntaxContext context)
+        {
+            if (context.IsStatementContext ||
+                context.IsGlobalStatementContext ||
+                context.IsPossibleTupleContext ||
+                context.IsPatternContext ||
```

Comment:
1. can you add a pattern test and a tuple test.  
2.  also a test that it doesn't appear here `enum e : $$`
3. what aobut here `nint n = new $$`
4. not after `as`
5. yes after `is`
6. after `fixed`?
7. in lambda parameter
8. `static operator implicit $$`
9. `<see cref="$$"/>`
10. after `ref`?
11. after const? (unless it's not legal there)
12. after `stackalloc`?
13. return type of a member in a class?

Reply:
need a test for a normal expression location.  like `var v = 1 + $$`.

Reply:
Great suggestions, thanks.

---
In reply to: [401371992](https://github.com/dotnet/roslyn/pull/42976#discussion_r401371992) [](ancestors = 401371992)

Reply:
> also a test that it doesn't appear here enum e : $$

I thought LDM agreed this was fine provided it didn't require unreasonable changes? I can't actually find the notes for that meeting however...

Reply:
> I thought LDM agreed this was fine provided it didn't require unreasonable changes? I can't actually find the notes for that meeting however...

If so, then add test that it's legal there :)

Reply:
Ah, here we go: https://github.com/dotnet/csharplang/blob/3b22aa0b47cc2ab6352e01f73f976fdbbf9d9b2d/meetings/2019/LDM-2019-10-23.md

> For use as an enum underlying type, we think nint/nuint should be allowed as long as it's not too much implementation work. The legal values would be the same as the valid constants for the underlying.

Reply:
I'll include the `enum` context when native integers are supported as underlying types in the compiler.

---
In reply to: [401782265](https://github.com/dotnet/roslyn/pull/42976#discussion_r401782265) [](ancestors = 401782265)